openmicroscopy/bioformats
# C++: Documentation, usability and tool consistency fixes

Overview:

- Move `showinf` and `glview` from `bin` to `libexec` (and rename to `info` and `view`, resp.)
- Add `bf` command to `bin`; this is the only user-visible executable
- Add manpages for `bf` and `bf-view`; update `bf-info`
- `--help` renamed to `--usage` (`bf` and `bf info`)
- `--help` now displays manual page (`bf` and `bf info`)

The reason for renaming `showinf` and `glview` is:

- to prevent future clashes with the Java implemenations both as commands and in the documentation (section names and crossref clashes)
- to provide a consistent and uniform interface to the end user
- moving them into `libexec` and invoking indirectly via `bf` removes these from the direct search path, and also makes it possible to replace the implementations transparently in the future

The `bf` command works as a front-end like the `git` command.  It's extensible, so simply dropping new scripts or binaries into the `libexec` dir will make them immediately available to the user.  This will ease the addition of more tools in updates.  The naming convention applies to the programs and the associated manual pages.  So "bf info" invokes libexec/info and has a `bf-info` manual page.  This is exactly the same as e.g. `git log` having a `git-log` manual page.  Its main purpose is to allow Bio-Formats C++ tools and libraries to be used without manual configuration of the user `PATH` or library search paths etcc; it handles all this for the user.  This will make it possible to download and run programs with zero set up overhead.

At the moment this only works on UNIX-like systems.  There's a stub cmd script for Windows for a future implementation equivalent to the `bf` script; the `--help` action will also need a Windows equivalent to display HTMLhelp or plain HTML sphinx docs in place of the manpage viewer.

--no-rebase

--------

Testing:

Build with the CMake option `-DCMAKE_INSTALL_PREFIX=` (no configured install path).  Build as normal, then install with `make DESTDIR=/tmp/c1 install`.

- Verify the presence of `/tmp/c1/bin/bf`, `/tmp/c1/libexec/info` and `/tmp/c1/libexec/view`

- Run (in the build tree) `./bf info --version` (should display version)
- Run (in the build tree) `./bf info --usage` (should display usage text)
- Run (in the build tree) `./bf info --help` (should display manual page)
- Run (in the build tree) `./bf info /path/to/ome.tiff` (should display image metadata)

- Run (anywhere) `/tmp/c1/bin/bf info --version` (should display version)
- Run (anywhere) `/tmp/c1/bin/bf info --usage` (should display usage text)
- Run (anywhere) `/tmp/c1/bin/bf info --help` (should display bf-info(1) manual page)
- Run (anywhere) `/tmp/c1/bin/bf info /path/to/ome.tiff` (should display image metadata)

- Run (anywhere) `/tmp/c1/bin/bf view` (should display window with toolbar icons)

- Run (anywhere) `/tmp/c1/bin/bf --version` (should display version)
- Run (anywhere) `/tmp/c1/bin/bf --usage` (should display usage text; this should show the `info` and `view` commands at the end)
- Run (anywhere) `/tmp/c1/bin/bf --help` (should display bf(1) manual page)
- Run (anywhere) `/tmp/c1/bin/bf bad` (should display error and nonzero exit status)
- Run (anywhere) `/tmp/c1/bin/bf` (should display usage text and nonzero exit status)

All `bf info` commands should also work using `bf showinf`, likewise `bf view` will work using `bf glview`.

This is testing that the PATH, [DY]LD_LIBRARY_PATH etc. are being set up correctly in the build and install trees, and that in the install tree, the runtime path introspection is working.  The `bf` script needs to set up the library search path in order to invoke the C++ binary, but after that the C++ code should be able to figure the rest out.

Build with the CMake option `-DCMAKE_INSTALL_PREFIX=/tmp/c2`, build as normal, then install with `make install`

- Verify the presence of `/tmp/c2/bin/bf`, `/tmp/c2/libexec/ome/bioformats/info` and `/tmp/c2/libexec/ome/bioformatsview`

- Run (in the build tree) `./bf info --version` (should display version)
- Run (in the build tree) `./bf info --usage` (should display usage text)
- Run (in the build tree) `./bf info --help` (should display manual page)
- Run (in the build tree) `./bf info /path/to/ome.tiff` (should display image metadata)

- Run (anywhere) `/tmp/c2/bin/bf info --version` (should display version)
- Run (anywhere) `/tmp/c2/bin/bf info --usage` (should display usage text)
- Run (anywhere) `/tmp/c2/bin/bf info --help` (should display manual page)
- Run (anywhere) `/tmp/c2/bin/bf info /path/to/ome.tiff` (should display image metadata)

- Run (anywhere) `/tmp/c2/bin/bf view` (should display window with toolbar icons)

- Run (anywhere) `/tmp/c2/bin/bf --version` (should display version)
- Run (anywhere) `/tmp/c2/bin/bf --usage` (should display usage text; this should show the `info` and `view` commands at the end)
- Run (anywhere) `/tmp/c2/bin/bf --help` (should display bf(1) manual page)
- Run (anywhere) `/tmp/c2/bin/bf bad` (should display error and nonzero exit status)
- Run (anywhere) `/tmp/c2/bin/bf` (should display usage text and nonzero exit status)

All `bf info` commands should also work using `bf showinf`, likewise `bf view` will work using `bf glview`.

This is testing that absolute paths work.  Other than needing a working library search path, no introspection should be needed here since it's explicitly configured (and non-relocatable).  The difference in libexec structure is to prevent clashes when installed into system locations such as /usr/local.
